Source/DafnyCore/DafnyOptions.cs

```diff
@@ -15,28 +19,87 @@
 using Bpl = Microsoft.Boogie;
 
 namespace Microsoft.Dafny {
+  public enum FunctionSyntaxOptions {
+    Version3,
+    Migration3To4,
+    ExperimentalTreatUnspecifiedAsGhost,
+    ExperimentalTreatUnspecifiedAsCompiled,
+    ExperimentalPredicateAlwaysGhost,
+    Version4,
+  }
+
+  public enum QuantifierSyntaxOptions {
+    Version3,
+    Version4,
+  }
 
-  public record Options(IDictionary<IOptionSpec, object> OptionArguments);
+  public record Options(IDictionary<Option, object> OptionArguments);
 
   public class DafnyOptions : Bpl.CommandLineOptions {
+
+    public void ApplyBinding(Option option) {
+      if (bindings.ContainsKey(option)) {
+        bindings[option](this, Get(option));
+      }
+    }
+
+    public T Get<T>(Option<T> option) {
+      return (T)Options.OptionArguments.GetOrCreate(option, () => default(T));
+    }
+
+    public object Get(Option option) {
+      return Options.OptionArguments[option];
+    }
+
+    public void Set(Option option, object value) {
+      Options.OptionArguments[option] = value;
+    }
+
+    public void Set<T>(Option<T> option, T value) {
+      Options.OptionArguments[option] = value;
+    }
+
     public void AddFile(string file) => base.AddFile(file, null);
 
-    private static readonly ISet<ILegacyOption> AvailableNewStyleOptions = new HashSet<ILegacyOption>(
-      new ILegacyOption[] {
-        ShowSnippetsOption.Instance,
-        TargetOption.Instance,
-        PrintOption.Instance,
-        LibraryOption.Instance,
-        PreludeOption.Instance,
-        PrintModeOption.Instance,
-        ResolvedPrintOption.Instance,
-        OldCompileVerboseOption.Instance,
-        OutputOption.Instance,
-        PluginOption.Instance,
-        UseRuntimeLibOption.Instance,
-        UnicodeCharactersOption.Instance,
-        OptimizeErasableDatatypeWrapperOption.Instance,
-      });
+    private static Dictionary<Option, Action<DafnyOptions, object>> bindings = new();
```

Comment:
```suggestion
    private static Dictionary<Option, Action<DafnyOptions, object>> legacyBindings = new();
```

Or are these used for anything other than legacy bindings?

Reply:
They're not.